---
title: "Snprelate"
output:
  html_document:
    dev: [svg, png]
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(SNPRelate)
library(tidyverse)
```

```{r metadata}
metadata = read.csv("https://github.com/borevitzlab/brachy-genotyping/raw/master/metadata/brachy-metadata.csv")
#str(metadata)
spp = read.csv("data/species_id.csv")
#str(spp)
```

```{r vcf2gds,cache=T}
#snpgdsVCF2GDS("data/2017-11-30_bhybridum_filtered_default.vcf.gz",
#              "data/2017-11-30_bhybridum_filtered_default.gds")
```

```{r geno.open}
geno = snpgdsOpen("data/2017-11-30_bhybridum_filtered_default.gds",
                  allow.duplicate = T, readonly = T)
```

```{r geno.summ,cache=T}
samp = snpgdsSummary(geno)$sample.id
```


## Filter out very bad missing data

```{r}
snp.missfilt = snpgdsSelectSNP(geno, missing.rate=0.999, autosome.only=F)
```

The total number of SNPs remaining is `r length(snp.missfilt)`.


## Species subset

Directly subset the three species' genotype matricies.                               

```{r samplesel}
chrom = read.gdsn(index.gdsn(geno, "snp.chromosome"))[snp.missfilt]
snp.dis = snp.missfilt[grep('Bd21-3', chrom)]
snp.sta = snp.missfilt[grep('Bstacei', chrom)]
snp.all = snp.missfilt
samp.hyb = filter(spp, species.id == "hybridum")$sample
samp.dis = filter(spp, species.id == "distachyon")$sample
samp.sta = filter(spp, species.id == "stacei")$sample
```


## Functions for further filtering

These things are in common for each species of brachy, so turn them into functions.

```{r}
ssp.filt = function(geno, samps, snps,  max.snp.miss.rate=0.99,
                    max.samp.miss.rate=0.99, min.maf=0.001 ) {
    miss.samp = snpgdsSampMissRate(geno, snp.id=snps, sample.id=samps)
    
    hist(miss.samp, breaks=100, main="Sample Missing Data (pre-filt)")
    abline(v=max.samp.miss.rate, col="blue", lwd=2)
    
    samps = samps[miss.samp <= max.samp.miss.rate]
    
    srf = snpgdsSNPRateFreq(geno, sample.id=samps, snp.id = snps)
    miss.snp = srf$MissingRate
    hist(miss.snp, breaks=100, main="SNP missing data")
    abline(v=max.snp.miss.rate, col="blue", lwd=2)
    
    maf = srf$MinorFreq
    hist(maf, breaks=50, main="SNP MAF")
    abline(v=min.maf, lwd=2, col="blue")
    
    snps = snpgdsSelectSNP(geno, sample.id=samps, snp.id=snps, maf=min.maf,
                                     missing.rate=max.snp.miss.rate, autosome.only=F)
    
    miss.samp = snpgdsSampMissRate(geno, snp.id=snps, sample.id=samps)
    hist(miss.samp, breaks=100, main="Sample Missing Data (post-filt)")
    
    print(paste("Num SNPs:", length(snps)))
    print(paste("Num Samples:", length(samps)))
    return(list(snps=snps, samps=samps, miss.samp=miss.samp))
}

ssp.geno = function(geno, filt) {
    ibs = snpgdsIBS(geno, sample.id=filt$samps, snp.id=filt$snps,
                    autosome.only=F, num.thread=4)
    ibs.nacnt = rowSums(is.na(ibs$ibs))
    table(ibs.nacnt)
    return(ibs)
}

distimpute = function(ibs, thresh=5, maxdist = 1) {
    dist = 1 - ibs$ibs 
    nasum = colSums(is.na(dist))
    pass = nasum < thresh
    dist = dist[pass,pass]
    
    ibs$sample.id = ibs$sample.id[pass]
    
    dist.ut = dist
    dist.ut[upper.tri(dist.ut,diag=T)] = 0
    nasum = colSums(is.na(dist.ut))
    
    # impute
    for (j in which(nasum > 0)) {
        for (i in which(is.na(dist[,j]))) {
            k = which(dist.ut[,j] == max(dist.ut[,j], na.rm=T))
            if (length(k) > 1) {
                k = k[1]
            }
            if (dist[k, j] <= maxdist) {
                dist[i, j] = max(dist[k, j], # k, j is neighbour -> self
                                 dist[k, i]) # k, i is neighbour -> other
                dist[j, i] = max(dist[k, j], # k, j is neighbour -> self
                                 dist[k, i]) # k, i is neighbour -> other
            }
        }
    }
    
    ibs$ibs = 1 - dist
    return(ibs)
}


distimpute2 = function(ibs, max.NAs=0, max.dist = 0.2) {
    dist = 1 - ibs$ibs 
    
    dist.ut = dist
    dist.ut[upper.tri(dist.ut,diag=T)] = 0
    nasum = colSums(is.na(dist.ut))
    num.imputed = 0
    # impute
    for (j in which(nasum > 0)) {
        for (i in which(is.na(dist[,j]))) {
            k = which(dist.ut[,j] == max(dist.ut[,j], na.rm=T))
            if (length(k) > 1) {
                k = k[1]
            }
            if (dist[k, j] <= max.dist) {
                num.imputed = num.imputed + 1
                dist[i, j] = max(dist[k, j], # k, j is neighbour -> self
                                 dist[k, i]) # k, i is neighbour -> other
                dist[j, i] = max(dist[k, j], # k, j is neighbour -> self
                                 dist[k, i]) # k, i is neighbour -> other
            }
        }
    }

    num.removed = 0
    while (sum(is.na(dist)) > 0) {
        rm = which.max(colSums(is.na(dist)))
        dist = dist[-rm,]
        dist = dist[,-rm]
        ibs$sample.id = ibs$sample.id[-rm]
        num.removed = num.removed + 1
    }
    ibs$ibs = 1 - dist
    ibs$num.imputed = num.imputed
    ibs$num.removed = num.removed
    print(paste("Num imputed:", num.imputed))
    print(paste("Num removed:", num.removed))
    return(ibs)
}
```


# Distachyon

```{r dis}
samp.rils = metadata %>%
    filter(grepl("^RIL", accession)) %>% 
    select(anon.name) %>%
    unlist
samp.dis.sansrils = samp.dis[! samp.dis %in% samp.rils]

filt.dis = ssp.filt(geno, samp.dis.sansrils, snp.dis, min.maf=0.01,
                    max.samp.miss.rate = 0.995, max.snp.miss.rate=0.95)

ibs.dis = ssp.geno(geno, filt.dis)
ibs.dis.imp = distimpute2(ibs.dis, max.dist = 0.2)
dim(ibs.dis.imp$ibs)
```

After the distance interpolation, we have interpolated `r ibs.dis.imp$num.imputed` entries, and have removed `r ibs.dis.imp$num.removed` samples.


## Find optimal z threshold

Run over a range of z thresholds.

```{r zthresh}
grouping = metadata %>% 
    select(accession, anon.name)

for (zthresh in c(seq(1, 4, 0.5), 5:10)) {
    cut.dis = snpgdsHCluster(ibs.dis.imp) %>% 
        snpgdsCutTree(outlier.n=0, z.threshold = zthresh)
    s = cut.dis$sample.id
    grouping = grouping %>% 
        filter(anon.name %in% cut.dis$sample.id)
    d = cut.dis$samp.group[match(grouping$anon.name, s)]
    snpgdsDrawTree(cut.dis)
    
    grouping[[paste0("GroupAt", zthresh, "_Ngrps", length(table(d)))]] = d
}

write.csv(grouping, "out/group_zscores.csv", row.names = F)
```

## Plot draft dendrogram

```{r bd-plot}
best.z.thresh = 2.5
ibs.dis.plt = ibs.dis.imp
ibs.dis.plt$sample.id = paste(metadata$anon.name[match(ibs.dis.imp$sample.id, metadata$anon.name)],
                              metadata$accession[match(ibs.dis.imp$sample.id, metadata$anon.name)])
                              

hc.dis.plt = snpgdsHCluster(ibs.dis.plt) %>% 
    snpgdsCutTree(outlier.n=1, z.threshold = best.z.thresh, label.H = F, label.Z = F)
snpgdsDrawTree(hc.dis.plt, leaflab="perpendicular")
```
```{r bdplot.pdf,include=F}
plotdendro = function (obj, clust.count = NULL, dend.idx = NULL, type = c("dendrogram", 
    "z-score"), yaxis.height = TRUE, yaxis.kinship = TRUE, y.kinship.baseline = NaN, 
    y.label.kinship = FALSE, outlier.n = NULL, shadow.col = c(rgb(0.5, 
        0.5, 0.5, 0.25), rgb(0.5, 0.5, 0.5, 0.05)), outlier.col = rgb(1, 
        0.5, 0.5, 0.5), leaflab = "none", labels = NULL, y.label = 0.2, 
    ...) 
{
    stopifnot(is.null(dend.idx) | is.numeric(dend.idx))
    type <- match.arg(type)
    stopifnot(is.numeric(y.kinship.baseline))
    if (type == "dendrogram") {
        stopifnot(!is.null(obj$dendrogram))
        stopifnot(is.null(outlier.n) | is.numeric(outlier.n))
        if (is.null(clust.count)) 
            clust.count <- obj$clust.count
        if (is.null(outlier.n)) 
            outlier.n <- obj$outlier.n
        if (!is.null(dend.idx)) {
            den <- obj$dendrogram[[dend.idx]]
            x.offset <- 0
            for (i in 1:length(dend.idx)) {
                if (dend.idx[i] == 2) {
                  IX <- dend.idx[1:i]
                  IX[i] <- 1
                  x.offset <- x.offset + attr(obj$dendrogram[[IX]], 
                    "member")
                }
            }
        }
        else {
            den <- obj$dendrogram
            x.offset <- 0
        }
        par(mar = c(8, 4, 4, 4))
        oldpar <- par(mgp = c(5, 1, 0))
        plot(den, leaflab = leaflab, axes = FALSE, ...)
        par(oldpar)
        if (yaxis.height) {
            axis(side = 2, line = 0)
            tmp <- list(...)
            if (!is.null(tmp$ylab)) 
                ylab <- tmp$ylab
            else ylab <- "individual dissimilarity"
            mtext(ylab, side = 2, line = 2.5)
        }
        if (yaxis.kinship) {
            if (is.finite(y.kinship.baseline)) {
                y.kinship.baseline <- y.kinship.baseline[1]
            }
            else {
                y.kinship.baseline <- attr(den, "height")
            }
            ym <- pretty(c(0, 1))
            axis(side = 4, (1 - ym) * y.kinship.baseline, ym, 
                line = 0)
            mtext("coancestry coefficient", 4, line = 2.5)
        }
        if (!is.null(clust.count)) {
            m <- c(0, cumsum(clust.count))
            jj <- 1
            k <- 1
            for (i in 1:length(clust.count)) {
                if (clust.count[i] > outlier.n) {
                  rect(m[i] + 0.5 - x.offset, par("usr")[3L], 
                    m[i + 1] + 0.5 - x.offset, par("usr")[4L], 
                    col = shadow.col[jj], border = NA)
                  jj <- 3 - jj
                  if (!is.null(labels[k])) 
                    text((m[i] + m[i + 1])/2 - x.offset, y.label, 
                      labels[k])
                  k <- k + 1
                }
                else {
                  rect(m[i] + 0.5 - x.offset, par("usr")[3L], 
                    m[i + 1] + 0.5 - x.offset, par("usr")[4L], 
                    col = outlier.col, border = NA)
                }
            }
        }
        if (0){#yaxis.kinship & y.label.kinship) {
            h1 <- (1 - 0.5) * y.kinship.baseline
            abline(h = h1, lty = 2, col = "gray")
            h2 <- (1 - 0.25) * y.kinship.baseline
            abline(h = h2, lty = 2, col = "gray")
            h3 <- (1 - 1/8) * y.kinship.baseline
            abline(h = h3, lty = 2, col = "gray")
            h4 <- (1 - 1/16) * y.kinship.baseline
            abline(h = h4, lty = 2, col = "gray")
            axis(side = 4, c(h1, h2, h3, h4), c("twins", "PC/FS", 
                "DFC/HS", "FC"), tick = FALSE, line = -0.75, 
                las = 2, cex.axis = 0.75, col.axis = "gray25")
        }
    }
    else if (type == "z-score") {
        if (is.null(obj$merge)) 
            stop("There is no Z score in this object.")
        y <- obj$merge[, 1]
        y <- y[order(y, decreasing = TRUE)]
        plot(y, xlab = "the order of Z score", ylab = "Z score", 
            type = "b", pch = "+", log = "x", ...)
        abline(h = 15, col = "gray", lty = 2)
    }
    invisible()
}

pdf(paste0("out/draft_dendro_bdis_z", best.z.thresh, ".pdf"), width=70, height=10)
par(mar=c(10,2, 1, 1))
plotdendro(hc.dis.plt, leaflab="perpendicular")
dev.off()
```

## Remove dodgy samples
```{r}
dodgy = read.csv("data/dodgy-samples.csv")$anon.name
samp.dis.genocall = samp.dis.sansrils[! samp.dis.sansrils %in% dodgy]

filt.dis.geno = ssp.filt(geno, samp.dis.genocall, snp.dis, min.maf=0.01,
                         max.samp.miss.rate = 0.995, max.snp.miss.rate=0.95)

ibs.dis = ssp.geno(geno, filt.dis.geno)
ibs.dis.imp = distimpute2(ibs.dis, max.dist = 0.2)
dim(ibs.dis.imp$ibs)
```

## Call genotypes

```{r best.z}
best.z.thresh = 2.5
best.z.hc = snpgdsHCluster(ibs.dis.imp) %>% 
    snpgdsCutTree(outlier.n=1, z.threshold = best.z.thresh, label.Z = F)
snpgdsDrawTree(best.z.hc)
sort(table(best.z.hc$samp.group), decreasing = T)
goodgroups = data.frame(best.z.hc[c("sample.id", "samp.group")]) %>% 
    filter(grepl("^G", samp.group))
```

```{r genocall}
genocall = metadata %>% 
    select(accession, anon.name) %>% 
    right_join(goodgroups, by=c("anon.name"="sample.id"))
genocall$missing.rate = snpgdsSampMissRate(geno, sample.id=as.character(genocall$anon.name),
                                           snp.id = filt.dis$snps)
write.csv(genocall, "out/genotypes.csv", row.names = F)
```

Now, we take the indiviudal run with the least missing data from each clonal lineage

```{r geno.best}
genocall.best = genocall %>% 
    group_by(samp.group) %>% 
    summarise(anon.name = anon.name[which.min(missing.rate)],
              accession = accession[which.min(missing.rate)],
              missing.rate = min(missing.rate))

write.csv(genocall.best, "out/genotypes_best.csv", row.names = F)
```

```{r filtersnps}
samps = genocall.best$anon.name
snps = filt.dis$snps
max.snp.miss.rate = 0.9
min.maf = 0.05

srf = snpgdsSNPRateFreq(geno, sample.id=samps, snp.id=snps)
miss.snp = srf$MissingRate
hist(miss.snp, breaks=100, main="SNP missing data")
abline(v=max.snp.miss.rate, lwd=2, col="blue")

maf = srf$MinorFreq
hist(maf, breaks=50, main="SNP MAF")
abline(v=min.maf, lwd=2, col="blue")

snps = snpgdsSelectSNP(geno, sample.id=samps, snp.id=snps, maf=min.maf,
                       missing.rate=max.snp.miss.rate, autosome.only=F)

srf.post = snpgdsSNPRateFreq(geno, sample.id=samps, snp.id=snps)
hist(srf.post$MissingRate, breaks=100, main="SNP missing data (post)")

print(paste("Num SNPs:", length(snps)))
print(paste("Num Samples:", length(samps)))
```

```{r structure}
struct.genos = snpgdsGetGeno(geno, sample.id = samps, snp.id=snps)
hist(colSums(struct.genos, na.rm = T))
table(colSums(struct.genos, na.rm = T)==2)
hist(colSums(is.na(struct.genos))/nrow(struct.genos))
rownames(struct.genos) = samps
struct.genos[struct.genos==1] = NA
write.table(struct.genos, "out/structure_genos_bdis.txt", na="-9",
            sep="\t", quote = F, col.names = F)
```

```{r bd-plotfinal}
ibs.dis.plt = ibs.dis.imp
ibs.dis.plt$sample.id = paste(metadata$anon.name[match(ibs.dis.imp$sample.id, metadata$anon.name)],
                              metadata$accession[match(ibs.dis.imp$sample.id, metadata$anon.name)])
                              

hc.dis.plt = snpgdsHCluster(ibs.dis.plt) %>% 
    snpgdsCutTree(outlier.n=1, z.threshold = best.z.thresh, label.H = F, label.Z = F)
snpgdsDrawTree(hc.dis.plt, leaflab="perpendicular")
```
```{r bdplot.pdf.final,include=F}
plotdendro = function (obj, clust.count = NULL, dend.idx = NULL, type = c("dendrogram", 
    "z-score"), yaxis.height = TRUE, yaxis.kinship = TRUE, y.kinship.baseline = NaN, 
    y.label.kinship = FALSE, outlier.n = NULL, shadow.col = c(rgb(0.5, 
        0.5, 0.5, 0.25), rgb(0.5, 0.5, 0.5, 0.05)), outlier.col = rgb(1, 
        0.5, 0.5, 0.5), leaflab = "none", labels = NULL, y.label = 0.2, 
    ...) 
{
    stopifnot(is.null(dend.idx) | is.numeric(dend.idx))
    type <- match.arg(type)
    stopifnot(is.numeric(y.kinship.baseline))
    if (type == "dendrogram") {
        stopifnot(!is.null(obj$dendrogram))
        stopifnot(is.null(outlier.n) | is.numeric(outlier.n))
        if (is.null(clust.count)) 
            clust.count <- obj$clust.count
        if (is.null(outlier.n)) 
            outlier.n <- obj$outlier.n
        if (!is.null(dend.idx)) {
            den <- obj$dendrogram[[dend.idx]]
            x.offset <- 0
            for (i in 1:length(dend.idx)) {
                if (dend.idx[i] == 2) {
                  IX <- dend.idx[1:i]
                  IX[i] <- 1
                  x.offset <- x.offset + attr(obj$dendrogram[[IX]], 
                    "member")
                }
            }
        }
        else {
            den <- obj$dendrogram
            x.offset <- 0
        }
        par(mar = c(8, 4, 4, 4))
        oldpar <- par(mgp = c(5, 1, 0))
        plot(den, leaflab = leaflab, axes = FALSE, ...)
        par(oldpar)
        if (yaxis.height) {
            axis(side = 2, line = 0)
            tmp <- list(...)
            if (!is.null(tmp$ylab)) 
                ylab <- tmp$ylab
            else ylab <- "individual dissimilarity"
            mtext(ylab, side = 2, line = 2.5)
        }
        if (yaxis.kinship) {
            if (is.finite(y.kinship.baseline)) {
                y.kinship.baseline <- y.kinship.baseline[1]
            }
            else {
                y.kinship.baseline <- attr(den, "height")
            }
            ym <- pretty(c(0, 1))
            axis(side = 4, (1 - ym) * y.kinship.baseline, ym, 
                line = 0)
            mtext("coancestry coefficient", 4, line = 2.5)
        }
        if (!is.null(clust.count)) {
            m <- c(0, cumsum(clust.count))
            jj <- 1
            k <- 1
            for (i in 1:length(clust.count)) {
                if (clust.count[i] > outlier.n) {
                  rect(m[i] + 0.5 - x.offset, par("usr")[3L], 
                    m[i + 1] + 0.5 - x.offset, par("usr")[4L], 
                    col = shadow.col[jj], border = NA)
                  jj <- 3 - jj
                  if (!is.null(labels[k])) 
                    text((m[i] + m[i + 1])/2 - x.offset, y.label, 
                      labels[k])
                  k <- k + 1
                }
                else {
                  rect(m[i] + 0.5 - x.offset, par("usr")[3L], 
                    m[i + 1] + 0.5 - x.offset, par("usr")[4L], 
                    col = outlier.col, border = NA)
                }
            }
        }
        if (0){#yaxis.kinship & y.label.kinship) {
            h1 <- (1 - 0.5) * y.kinship.baseline
            abline(h = h1, lty = 2, col = "gray")
            h2 <- (1 - 0.25) * y.kinship.baseline
            abline(h = h2, lty = 2, col = "gray")
            h3 <- (1 - 1/8) * y.kinship.baseline
            abline(h = h3, lty = 2, col = "gray")
            h4 <- (1 - 1/16) * y.kinship.baseline
            abline(h = h4, lty = 2, col = "gray")
            axis(side = 4, c(h1, h2, h3, h4), c("twins", "PC/FS", 
                "DFC/HS", "FC"), tick = FALSE, line = -0.75, 
                las = 2, cex.axis = 0.75, col.axis = "gray25")
        }
    }
    else if (type == "z-score") {
        if (is.null(obj$merge)) 
            stop("There is no Z score in this object.")
        y <- obj$merge[, 1]
        y <- y[order(y, decreasing = TRUE)]
        plot(y, xlab = "the order of Z score", ylab = "Z score", 
            type = "b", pch = "+", log = "x", ...)
        abline(h = 15, col = "gray", lty = 2)
    }
    invisible()
}

pdf(paste0("out/final_dendro_bdis_z", best.z.thresh, ".pdf"), width=70, height=10)
par(mar=c(10,2, 1, 1))
plotdendro(hc.dis.plt, leaflab="perpendicular")
dev.off()
```

# Hybridum and stacei

```{r hyb}
filt.hyb = ssp.filt(geno, samp.hyb, snp.all, min.maf=0.01,
                    max.samp.miss.rate = 0.99, max.snp.miss.rate=0.97)

ibs.hyb = ssp.geno(geno, filt.hyb)
ibs.hyb.imp = distimpute2(ibs.hyb, max.dist = 0.2)

table(colSums(is.na(ibs.hyb.imp$ibs)))
snpgdsHCluster(ibs.hyb.imp) %>% 
    snpgdsCutTree() %>% 
    snpgdsDrawTree(outlier.n = 0)
```


```{r sta}
filt.sta = ssp.filt(geno, samp.sta, snp.sta, min.maf=0.01,
                    max.samp.miss.rate = 0.99, max.snp.miss.rate=0.97)
ibs.sta = ssp.geno(geno, filt.sta)
ibs.sta.imp = distimpute2(ibs.hyb, max.dist = 0.2)

table(colSums(is.na(ibs.sta.imp$ibs)))
snpgdsHCluster(ibs.sta.imp) %>% 
    snpgdsCutTree(outlier.n=0) %>% 
    snpgdsDrawTree()
```